diff --git a/torchrec_dlrm/data/dlrm_dataloader.py b/torchrec_dlrm/data/dlrm_dataloader.py
index 4bb9001..ca58ddb 100644
--- a/torchrec_dlrm/data/dlrm_dataloader.py
+++ b/torchrec_dlrm/data/dlrm_dataloader.py
@@ -130,7 +130,7 @@ def _get_in_memory_dataloader(
             ),
         ),
         batch_size=None,
-        pin_memory=args.pin_memory,
+        pin_memory=False,
         collate_fn=lambda x: x,
     )
     return dataloader
diff --git a/torchrec_dlrm/dlrm_main.py b/torchrec_dlrm/dlrm_main.py
index a3d98b6..52b287e 100644
--- a/torchrec_dlrm/dlrm_main.py
+++ b/torchrec_dlrm/dlrm_main.py
@@ -25,7 +25,7 @@ from torchrec.distributed.model_parallel import (
     DistributedModelParallel,
     get_default_sharders,
 )
-from torchrec.distributed.planner import EmbeddingShardingPlanner, Topology
+from torchrec.distributed.planner import EmbeddingShardingPlanner, Topology, ParameterConstraints
 from torchrec.distributed.planner.storage_reservations import (
     HeuristicalStorageReservation,
 )
@@ -36,6 +36,13 @@ from torchrec.optim.keyed import CombinedOptimizer, KeyedOptimizerWrapper
 from torchrec.optim.optimizers import in_backward_optimizer_filter
 from tqdm import tqdm
 
+from torchrec.distributed.types import ShardingEnv
+from hybrid_torchrec import HashEmbeddingBagCollection, HashEmbeddingBagConfig
+from hybrid_torchrec.distributed.sharding_plan import get_default_hybrid_sharders
+from hybrid_torchrec.distributed.hybrid_train_pipeline import (
+    HybridTrainPipelineSparseDist,
+)
+
 # OSS import
 try:
     # pyre-ignore[21]
@@ -334,7 +341,7 @@ def _evaluate(
 
     iterator = itertools.islice(iter(eval_dataloader), limit_batches)
 
-    auroc = metrics.AUROC(compute_on_step=False, num_classes=2).to(device)
+    auroc = metrics.AUROC("binary").to(device)
 
     is_rank_zero = dist.get_rank() == 0
     if is_rank_zero:
@@ -472,7 +479,7 @@ def train_val_test(
         TrainValTestResults.
     """
     results = TrainValTestResults()
-    pipeline = TrainPipelineSparseDist(
+    pipeline = HybridTrainPipelineSparseDist(
         model, optimizer, device, execute_all_batches=True
     )
 
@@ -520,7 +527,8 @@ def main(argv: List[str]) -> None:
         except (ValueError, AttributeError):
             pass
 
-    torch.backends.cuda.matmul.allow_tf32 = args.allow_tf32
+    import torch_npu
+    torch_npu.npu.matmul.allow_hf32 = args.allow_tf32
 
     if args.multi_hot_sizes is not None:
         assert (
@@ -546,6 +554,10 @@ def main(argv: List[str]) -> None:
         device: torch.device = torch.device(f"cuda:{rank}")
         backend = "nccl"
         torch.cuda.set_device(device)
+    elif torch_npu.npu.is_available():
+        device: torch.device = torch.device(f"npu:{rank}")
+        backend = "hccl"
+        torch_npu.npu.set_device(device)
     else:
         device: torch.device = torch.device("cpu")
         backend = "gloo"
@@ -594,7 +606,7 @@ def main(argv: List[str]) -> None:
 
     if args.interaction_type == InteractionType.ORIGINAL:
         dlrm_model = DLRM(
-            embedding_bag_collection=EmbeddingBagCollection(
+            embedding_bag_collection=HashEmbeddingBagCollection(
                 tables=eb_configs, device=torch.device("meta")
             ),
             dense_in_features=len(DEFAULT_INT_NAMES),
@@ -604,7 +616,7 @@ def main(argv: List[str]) -> None:
         )
     elif args.interaction_type == InteractionType.DCN:
         dlrm_model = DLRM_DCN(
-            embedding_bag_collection=EmbeddingBagCollection(
+            embedding_bag_collection=HashEmbeddingBagCollection(
                 tables=eb_configs, device=torch.device("meta")
             ),
             dense_in_features=len(DEFAULT_INT_NAMES),
@@ -616,7 +628,7 @@ def main(argv: List[str]) -> None:
         )
     elif args.interaction_type == InteractionType.PROJECTION:
         dlrm_model = DLRM_Projection(
-            embedding_bag_collection=EmbeddingBagCollection(
+            embedding_bag_collection=HashEmbeddingBagCollection(
                 tables=eb_configs, device=torch.device("meta")
             ),
             dense_in_features=len(DEFAULT_INT_NAMES),
@@ -631,6 +643,9 @@ def main(argv: List[str]) -> None:
             "Unknown interaction option set. Should be original, dcn, or projection."
         )
 
+    host_gp = dist.new_group(backend='gloo')
+    host_env = ShardingEnv(world_size=dist.get_world_size(), rank=rank, pg=host_gp)
+
     train_model = DLRMTrain(dlrm_model)
     embedding_optimizer = torch.optim.Adagrad if args.adagrad else torch.optim.SGD
     # This will apply the Adagrad optimizer in the backward pass for the embeddings (sparse_arch). This means that
@@ -648,23 +663,29 @@ def main(argv: List[str]) -> None:
         train_model.model.sparse_arch.parameters(),
         optimizer_kwargs,
     )
+    constraints = {
+        f"t_{feature_name}": ParameterConstraints(
+            sharding_types=["row_wise"],
+            compute_kernels=["fused"]
+        )
+        for feature_idx, feature_name in enumerate(DEFAULT_CAT_NAMES)
+    }
+    
     planner = EmbeddingShardingPlanner(
         topology=Topology(
             local_world_size=get_local_size(),
             world_size=dist.get_world_size(),
             compute_device=device.type,
         ),
-        batch_size=args.batch_size,
-        # If experience OOM, increase the percentage. see
-        # https://pytorch.org/torchrec/torchrec.distributed.planner.html#torchrec.distributed.planner.storage_reservations.HeuristicalStorageReservation
-        storage_reservation=HeuristicalStorageReservation(percentage=0.05),
+        constraints=constraints,
     )
     plan = planner.collective_plan(
-        train_model, get_default_sharders(), dist.GroupMember.WORLD
+        train_model, get_default_hybrid_sharders(host_env), dist.GroupMember.WORLD
     )
 
     model = DistributedModelParallel(
         module=train_model,
+        sharders=get_default_hybrid_sharders(host_env),
         device=device,
         plan=plan,
     )
diff --git a/torchrec_dlrm/multi_hot.py b/torchrec_dlrm/multi_hot.py
index 77b13b9..7ca0d24 100644
--- a/torchrec_dlrm/multi_hot.py
+++ b/torchrec_dlrm/multi_hot.py
@@ -19,7 +19,7 @@ class RestartableMap:
 
     def __iter__(self):
         for x in self.source:
-            yield self.func(x)
+            yield self.func(x).pin_memory()
 
     def __len__(self):
         return len(self.source)
@@ -108,7 +108,7 @@ class Multihot:
             )
             multi_hot_tables_l.append(multi_hot_table)
         multi_hot_tables_l = [
-            torch.from_numpy(multi_hot_table).int()
+            torch.from_numpy(multi_hot_table).long()
             for multi_hot_table in multi_hot_tables_l
         ]
         return multi_hot_tables_l
